python/cudf/cudf/core/frame.py

```diff
@@ -438,128 +438,244 @@ def _normalize_columns_and_scalars_type(self, other):
-                self._column, other, boolean_mask._column
-            )
+
+            result = libcudf.copying.copy_if_else(input_col, other, cond)
 
             if is_categorical_dtype(self.dtype):
                 result = build_categorical_column(
-                    categories=self._column.categories,
+                    categories=self._data[self.name].categories,
                     codes=as_column(result.base_data, dtype=result.dtype),
                     mask=result.base_mask,
                     size=result.size,
                     offset=result.offset,
-                    ordered=self._column.ordered,
+                    ordered=self._data[self.name].ordered,
                 )
 
-            result = self.__class__(result)
-            result.index = self.index
-            return result
+            if isinstance(self, cudf.Index):
+                from cudf.core.index import as_index
+
+                result = as_index(result, name=self.name)
+            else:
+                result = self._copy_construct(data=result)
+
+            return self._mimic_inplace(result, inplace=inplace)
+
+    def mask(self, cond, other=None, inplace=False):
+        """
+        Replace values where the condition is True.
+
+        Parameters
+        ----------
+        cond : bool Series/DataFrame, array-like
+            Where cond is False, keep the original value.
+            Where True, replace with corresponding value from other.
+            Callables are not supported.
+        other: scalar, list of scalars, Series/DataFrame
+            Entries where cond is True are replaced with
+            corresponding value from other. Callables are not
+            supported. Default is None.
+
+            DataFrame expects only Scalar or array like with scalars or
+            dataframe with same dimention as self.
+
+            Series expects only scalar or series like with same length
+        inplace : bool, default False
+            Whether to perform the operation in place on the data.
+
+        Returns
+        -------
+        Same type as caller
+
+        Examples:
+        ---------
+        >>> import cudf
+        >>> df = cudf.DataFrame({"A":[1, 4, 5], "B":[3, 5, 8]})
+        >>> df.mask(df % 2 == 0, [-1, -1])
+           A  B
+        0  1  3
+        1 -1  5
+        2  5 -1
+
+        >>> ser = cudf.Series([4, 3, 2, 1, 0])
+        >>> ser.mask(ser > 2, 10)
+        0    10
+        1    10
+        2     2
+        3     1
+        4     0
+        dtype: int64
+        >>> ser.mask(ser > 2)
+        0    null
+        1    null
+        2       2
+        3       1
+        4       0
+        dtype: int64
+        """
+
+        if not hasattr(cond, "__invert__"):
+            # We Invert `cond` below and call `where`, so
+            # making sure the object supports
+            # `~`(inversion) operator or `__invert__` method
+            cond = cupy.array(cond)
```

Comment:
This allows us to handle GPU objects zero copy:
```suggestion
            cond = cupy.asarray(cond)
```